File: generator/src/main/java/ch/bailu/gtk/converter/ReservedTokenTable.java
package ch.bailu.gtk.converter;

import java.util.HashMap;
import java.util.Map;

public class ReservedTokenTable {
    
    private final Map<String, String> table = new HashMap<>();

    private static ReservedTokenTable INSTANCE = null;

    public static ReservedTokenTable instance() {
        if (INSTANCE == null) {
            INSTANCE = new ReservedTokenTable();
        }
        return INSTANCE;
    }

    private ReservedTokenTable() {
        add("native",   "xnative");
        add("continue", "xcontinue");
        add("double",   "xdouble");
        add("int",      "xint");
        add("new",      "xnew");
        add("default",  "xdefault");
        add("private",  "xprivate");
        add("...",      "xelipse");
        add("notify",   "xnotify");
        add("interface","xinterface");
        add("2BUTTON_PRESS", "TWO_BUTTON_PRESS");
        add("3BUTTON_PRESS", "TREE_BUTTON_PRESS");
        add("2BIG", "_2_BIG");
        add("false", "FALSE");
        add("true", "TRUE");
        add("ch",   "CH");
        add("toString", "toStr");
    }


    private void add(String reserved, String replacement) {
        table.put(reserved, replacement);
    }

    public String convert(String token) {
        String result =  table.get(token);
        if (result == null) {
            result = token;
        }
        return result;
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/table/ReservedTokenTable.kt
package ch.bailu.gtk.table

object ReservedTokenTable {
    private val table: MutableMap<String, String> = HashMap()

    init {
        add("package", "_package")
        add("native", "_native")
        add("continue", "_continue")
        add("double", "_double")
        add("int", "_int")
        add("new", "_new")
        add("default", "_default")
        add("private", "_private")
        add("...", "_ellipsis")
        add("interface", "_interface")
        add("false", "_false")
        add("true", "_true")
        add("toString", "toStr")
        add("wait", "doWait")
    }

    private fun add(reserved: String, replacement: String) {
        table[reserved] = replacement
    }

    fun convert(token: String): String {
        return table[token] ?: convertNumber(token)
    }

    private fun convertNumber(token: String): String {
        if (token.isNotEmpty() && token[0].isDigit()) {
            return "_$token"
        }
        return token
    }
}
